volumina/utility/preferences.py

```diff
@@ -25,75 +25,73 @@
+        self.path = path
+        self._lock = threading.RLock()
+
+    @property
+    def path(self):
+        return self._path
+
+    @path.setter
+    def path(self, path: Union[str, bytes, os.PathLike]):
+        self._path = pathlib.Path(path)
 
     def get(self, group: str, setting: str, default: Any = None) -> Any:
         return self.getmany((group, setting, default))[0]
 
     def set(self, group: str, setting: str, value: Any) -> None:
         self.setmany((group, setting, value))
 
-    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple:
         with self._lock:
+            data = self._read()
             result = []
             for group, setting, default in args:
                 try:
-                    result.append(self._data[group][setting])
+                    result.append(data[group][setting])
                 except KeyError:
                     result.append(default)
             return tuple(result)
 
     def setmany(self, *args: Tuple[str, str, Any]) -> None:
         with self._lock:
+            data = self._read()
             for group, setting, value in args:
-                self._data.setdefault(group, {})[setting] = value
-            try:
-                with open(self._path, "wb") as f:
-                    # Use the lowest available protocol for backwards compatibility.
-                    pickle.dump(self._data, f, protocol=0)
-            except Exception as e:
-                logger.exception(f"Failed to save preferences to {str(self._path)!r}")
-                if not isinstance(e, IOError):
-                    self._path.unlink()
-
-    def get_location(self) -> pathlib.Path:
-        with self._lock:
-            return self._path
-
-    def set_location(self, path: os.PathLike) -> None:
-        with self._lock:
-            self._path = pathlib.Path(path)
-            self._data = _load_preferences(self._path)
+                data.setdefault(group, {})[setting] = value
+            self._write(data)
 
+    def _read(self):
+        try:
+            with open(self.path) as f:
+                return json.load(f)
+        except FileNotFoundError:
+            return {}
 
-def _load_preferences(path: pathlib.Path) -> MutableMapping[str, MutableMapping[str, Any]]:
-    try:
-        with open(path, "rb") as f:
-            return pickle.load(f)
-    except FileNotFoundError:
-        return {}
-    except Exception as e:
-        logger.exception(f"Failed to load preferences from {str(path)!r}")
-        if not isinstance(e, IOError):
-            path.unlink()
-        return {}
+    def _write(self, data):
+        with open(self.path, "w") as f:
+            json.dump(data, f, indent=2, sort_keys=True)
 
 
-_preferences = _Preferences(pathlib.Path.home() / ".ilastik_preferences")
+_preferences = Preferences(
+    os.path.join(
+        appdirs.user_config_dir(appname="ilastik", appauthor=False),
```

Comment:
I see you're moving the config away from just home. It is good to "respect" the OS's wishes here. But this also comes with the downside that you cannot simply point users to their home folders, but need to remember (check) where the specific config location are on the specific OS. So I'd rather keep it in home.

Reply:
The better approach may be adding a menu item "Reveal in (Finder|Explorer|etc.)" to ilastik GUI that opens a config folder.

Reply:
We can also use this folder to keep other ilastik stuff.

Reply:
> The better approach may be adding a menu item "Reveal in (Finder|Explorer|etc.)" to ilastik GUI that opens a config folder.

Just a story from the past: ilastik wouldn't show up for users, thought it clearly started. Turned out geometry was off, so the window was spawned somewhere in nirvana. Deleting the preferences solved the issue. A "reveal in whatever" function wouldn't have been helpful there.

Reply:
but yeah, this is a fringe case. They certainly look better somewhere not in `$HOME`, since it is not something the user should play around with anyway.

Reply:
> Just a story from the past: ...

That's a good point!

Reply:
So, what should we do? I still think it's nice to keep configs and other data in the proper directories, and tell users in the documentation where to look if they stumble upon some obscure bug.

Reply:
nah, that's fine :)